test/cpp/api/serialize.cpp

```diff
@@ -271,6 +458,48 @@ TEST(SerializeTest, Optim) {
   }
 }
 
+TEST(SerializeTest, Optim_Adagrad) {
+  test_serialize_optimizer<Adagrad, AdagradOptions, AdagradParamState>(AdagradOptions(1e-1));
+
+  // bc compatibility check
+  auto model1 = Linear(5, 2);
+  auto optim1 = torch::optim::Adagrad(
+      model1->parameters(), torch::optim::AdagradOptions(1e-1));
+
+  auto x = torch::ones({10, 5});
+  auto step = [&x](torch::optim::Optimizer& optimizer, Linear model) {
+    optimizer.zero_grad();
+    auto y = model->forward(x).sum();
+    y.backward();
+    optimizer.step();
+  };
+  step(optim1, model1);
+  auto optim1_2 = Adagrad(model1->parameters(), torch::optim::AdagradOptions(1e-1));
+
+  std::vector<torch::Tensor> sum_buffers; // fill up with optim1 sum_buffers
+  std::vector<int64_t> step_buffers; // fill up with optim1 state_buffers
+  const auto& params_ = optim1.param_groups()[0].params();
+  const auto& optim1_state = optim1.state();
+  for (size_t i = 0; i < params_.size(); i++) {
+    auto key_ = c10::guts::to_string(params_[i].unsafeGetTensorImpl());
+    const AdagradParamState& curr_state_ = static_cast<const AdagradParamState&>(*(optim1_state.at(key_).get()));
+    sum_buffers.emplace_back(curr_state_.sum());
+    step_buffers.emplace_back(curr_state_.step());
+  }
+  // write sum_buffers and step_buffers to the file
+  auto optim_tempfile_old_format = c10::make_tempfile();
+  torch::serialize::OutputArchive output_archive;
+  write_tensors_to_archive(output_archive, "sum_buffers", sum_buffers);
+  write_step_buffers(output_archive, "step_buffers", step_buffers);
+  output_archive.save_to(optim_tempfile_old_format.name);
+  torch::serialize::InputArchive input_archive;
+  //OLD_SERIALIZATION_LOGIC_WARNING_CHECK("Adagrad", input_archive, load_from, optim_tempfile_old_format.name);
```

Comment:
@yf225 this is failing for some reason I don't know why. can you look at the macro and see if it's okay or why it might be failing

Reply:
Would writing it as `OLD_SERIALIZATION_LOGIC_WARNING_CHECK(Adagrad, input_archive, load_from, optim_tempfile_old_format.name)` and use `#optimizer_name` in the macro implementation work?

Reply:
Also I am curious what build error did you see?